File: android/src/main/java/com/ReactNativeBlobUtil/ReactNativeBlobUtilUtils.java
package com.ReactNativeBlobUtil;

import android.net.Uri;
import android.util.Base64;

import com.ReactNativeBlobUtil.Utils.PathResolver;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import android.content.Context;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.Charset;
import java.security.KeyStore;
import java.security.MessageDigest;
import java.security.cert.Certificate;
import java.security.cert.CertificateFactory;
import java.util.List;
import java.util.Locale;

import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSession;
import javax.net.ssl.SSLSocketFactory;
import javax.net.ssl.TrustManager;
import javax.net.ssl.TrustManagerFactory;
import javax.net.ssl.X509TrustManager;

import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;

public class ReactNativeBlobUtilUtils {

    public static X509TrustManager sharedTrustManager;

    public static String getMD5(String input) {
        String result = null;

        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            md.update(input.getBytes());
            byte[] digest = md.digest();

            StringBuilder sb = new StringBuilder();

            for (byte b : digest) {
                sb.append(String.format(Locale.ROOT, "%02x", b & 0xff));
            }

            result = sb.toString();
        } catch (Exception ex) {
            ex.printStackTrace();
        } finally {
            // TODO: Is discarding errors the intent? (https://www.owasp.org/index.php/Return_Inside_Finally_Block)
            return result;
        }

    }

    public static void emitWarningEvent(String data) {
        WritableMap args = Arguments.createMap();
        args.putString("event", "warn");
        args.putString("detail", data);

        // emit event to js context
        ReactNativeBlobUtilImpl.RCTContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(ReactNativeBlobUtilConst.EVENT_MESSAGE, args);
    }

    public static OkHttpClient.Builder getUnsafeOkHttpClient(OkHttpClient client) {
        try {

            if (sharedTrustManager == null) throw new IllegalStateException("Use of own trust manager but none defined");

            final TrustManager[] trustAllCerts = new TrustManager[]{sharedTrustManager};

            // Install the all-trusting trust manager
            final SSLContext sslContext = SSLContext.getInstance("SSL");
            sslContext.init(null, trustAllCerts, new java.security.SecureRandom());
            // Create an ssl socLket factory with our all-trusting manager
            final SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();

            OkHttpClient.Builder builder = client.newBuilder();
            builder.sslSocketFactory(sslSocketFactory, sharedTrustManager);
            builder.hostnameVerifier(new HostnameVerifier() {
                @Override
                public boolean verify(String hostname, SSLSession session) {
                    return true;
                }
            });

            return builder;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Whether a request to {@code url} should use the custom CA trust store.
     *
     * When pinnedHosts is set the custom CA only covers those hosts; anything else
     * falls through to the platform trust store, which is what the README documents
     * and what iOS does by returning NSURLSessionAuthChallengePerformDefaultHandling.
     */
    public static boolean customCACertsApplyTo(List<String> certNames, List<String> pinnedHosts, String url) {
        if (certNames == null || certNames.isEmpty()) return false;
        if (pinnedHosts == null || pinnedHosts.isEmpty()) return true;

        HttpUrl parsed = HttpUrl.parse(url);
        if (parsed == null) return false;

        return pinnedHosts.contains(parsed.host());
    }

    public static OkHttpClient.Builder getCustomCACertOkHttpClient(OkHttpClient client, Context context, List<String> certNames, boolean trustSystemCerts) {
        try {
            CertificateFactory cf = CertificateFactory.getInstance("X.509");
            KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
            keyStore.load(null, null);

            if (trustSystemCerts) {
                TrustManagerFactory defaultTmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
                defaultTmf.init((KeyStore) null);
                int systemIndex = 0;
                for (TrustManager tm : defaultTmf.getTrustManagers()) {
                    if (tm instanceof X509TrustManager) {
                        for (java.security.cert.X509Certificate cert : ((X509TrustManager) tm).getAcceptedIssuers()) {
                            keyStore.setCertificateEntry("system_" + (systemIndex++), cert);
                        }
                    }
                }
            }

            int loaded = 0;
            for (String certName : certNames) {
                Certificate cert = loadCertificateFromResources(context, cf, certName);
                if (cert != null) {
                    keyStore.setCertificateEntry(certName, cert);
                    loaded++;
                }
            }

            // Fail closed. Without this the keystore is empty, every connection fails
            // trust evaluation, and the developer sees a generic handshake error rather
            // than the actual problem: the certificate is not in res/raw under that name.
            if (loaded == 0) {
                throw new IllegalStateException(
                    "customCACerts: none of " + certNames + " could be loaded from res/raw");
            }

            TrustManagerFactory tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
            tmf.init(keyStore);

            X509TrustManager customTrustManager = null;
            for (TrustManager tm : tmf.getTrustManagers()) {
                if (tm instanceof X509TrustManager) {
                    customTrustManager = (X509TrustManager) tm;
                    break;
                }
            }

            if (customTrustManager == null) {
                throw new IllegalStateException("No X509TrustManager found");
            }

            SSLContext sslContext = SSLContext.getInstance("TLS");
            sslContext.init(null, new TrustManager[]{customTrustManager}, new java.security.SecureRandom());

            OkHttpClient.Builder builder = client.newBuilder();
            builder.sslSocketFactory(sslContext.getSocketFactory(), customTrustManager);

            // No hostnameVerifier override: the default verifier still applies, so a
            // certificate is only accepted for the names it actually carries. Scoping to
            // pinnedHosts happens before this client is built - see ReactNativeBlobUtilReq.

            return builder;
        } catch (Exception e) {
            throw new RuntimeException("Failed to configure custom CA certificates", e);
        }
    }

    private static Certificate loadCertificateFromResources(Context context, CertificateFactory cf, String certName) {
        int resId = context.getResources().getIdentifier(certName, "raw", context.getPackageName());
        if (resId == 0) {
            String[] suffixes = {"_pem", "_der", "_cer"};
            for (String suffix : suffixes) {
                resId = context.getResources().getIdentifier(certName + suffix, "raw", context.getPackageName());
                if (resId != 0) break;
            }
        }
        if (resId == 0) return null;

        try (InputStream is = context.getResources().openRawResource(resId)) {
            return cf.generateCertificate(is);
        } catch (Exception e) {
            // DER parsing failed — try PEM-to-DER conversion
        }

        try (InputStream is = context.getResources().openRawResource(resId)) {
            byte[] derBytes = pemToDer(is);
            if (derBytes != null) {
                return cf.generateCertificate(new ByteArrayInputStream(derBytes));
            }
        } catch (Exception e) {
            // PEM conversion also failed
        }

        return null;
    }

    private static byte[] pemToDer(InputStream is) {
        try {
            BufferedReader reader = new BufferedReader(new InputStreamReader(is));
            StringBuilder base64 = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                if (line.startsWith("-----")) continue;
                base64.append(line.trim());
            }
            return Base64.decode(base64.toString(), Base64.DEFAULT);
        } catch (Exception e) {
            return null;
        }
    }

    /**
     * String to byte converter method
     *
     * @param data     Raw data in string format
     * @param encoding Decoder name
     * @return Converted data byte array
     */
    public static byte[] stringToBytes(String data, String encoding) {
        if (encoding.equalsIgnoreCase("ascii")) {
            return data.getBytes(Charset.forName("US-ASCII"));
        } else if (encoding.toLowerCase(Locale.ROOT).contains("base64")) {
            return Base64.decode(data, Base64.NO_WRAP);

        } else if (encoding.equalsIgnoreCase("utf8")) {
            return data.getBytes(Charset.forName("UTF-8"));
        }
        return data.getBytes(Charset.forName("US-ASCII"));
    }

    /**
     * Normalize the path, remove URI scheme (xxx://) so that we can handle it.
     *
     * @param path URI string.
     * @return Normalized string
     */
    public static String normalizePath(String path) {
        if (path == null)
            return null;
        if (!path.matches("\\w+\\:.*"))
            return path;
        if (path.startsWith("file://")) {
            return path.replace("file://", "");
        }

        Uri uri = Uri.parse(path);
        if (path.startsWith(ReactNativeBlobUtilConst.FILE_PREFIX_BUNDLE_ASSET)) {
            return path;
        } else
            return PathResolver.getRealPathFromURI(ReactNativeBlobUtilImpl.RCTContext, uri);
    }

    public static boolean isAsset(String path) {
        return path != null && path.startsWith(ReactNativeBlobUtilConst.FILE_PREFIX_BUNDLE_ASSET);
    }

    public static boolean isContentUri(String path) {
        return path != null && path.startsWith(ReactNativeBlobUtilConst.FILE_PREFIX_CONTENT);
    }
}

File: android/src/main/java/com/ReactNativeBlobUtil/ReactNativeBlobUtilUtils.kt
package com.ReactNativeBlobUtil

import android.content.Context
import android.util.Base64
import com.facebook.react.bridge.Arguments
import com.facebook.react.modules.core.DeviceEventManagerModule
import okhttp3.OkHttpClient
import java.io.BufferedReader
import java.io.ByteArrayInputStream
import java.io.InputStream
import java.io.InputStreamReader
import java.nio.charset.Charset
import java.security.KeyStore
import java.security.MessageDigest
import java.security.SecureRandom
import java.security.cert.Certificate
import java.security.cert.CertificateFactory
import java.util.Locale
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager
import javax.net.ssl.TrustManagerFactory
import javax.net.ssl.X509TrustManager

/**
 * A class with a companion rather than an object, so the public members keep the
 * shape apps already use from Java and Kotlin: `ReactNativeBlobUtilUtils.sharedTrustManager = ...`.
 */
class ReactNativeBlobUtilUtils {

    companion object {

        /**
         * Trust manager for requests made with `trusty: true`. The library ships
         * none; an app that needs trusty sets its own (see the README).
         */
        @JvmField
        var sharedTrustManager: X509TrustManager? = null

        private val SCHEME = Regex("\\w+\\:.*")

        @JvmStatic
        fun getMD5(input: String?): String? {
            if (input == null) {
                return null
            }
            return try {
                val md = MessageDigest.getInstance("MD5")
                md.update(input.toByteArray())
                md.digest().joinToString("") { String.format(Locale.ROOT, "%02x", it.toInt() and 0xff) }
            } catch (ex: Exception) {
                ex.printStackTrace()
                null
            }
        }

        @JvmStatic
        fun emitWarningEvent(data: String?) {
            val args = Arguments.createMap()
            args.putString("event", "warn")
            args.putString("detail", data)

            // emit event to js context
            ReactNativeBlobUtilImpl.RCTContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit(ReactNativeBlobUtilConst.EVENT_MESSAGE, args)
        }

        @JvmStatic
        fun getUnsafeOkHttpClient(client: OkHttpClient): OkHttpClient.Builder {
            try {
                val trustManager = sharedTrustManager
                    ?: throw IllegalStateException("Use of own trust manager but none defined")

                val trustAllCerts = arrayOf<TrustManager>(trustManager)

                // Install the all-trusting trust manager
                val sslContext = SSLContext.getInstance("SSL")
                sslContext.init(null, trustAllCerts, SecureRandom())
                // Create an ssl socket factory with our all-trusting manager
                val sslSocketFactory = sslContext.socketFactory

                val builder = client.newBuilder()
                builder.sslSocketFactory(sslSocketFactory, trustManager)
                // The hostname is still verified. trusty used to switch that off too, so
                // an app whose trust manager validated a private CA properly still
                // accepted that CA's certificate for any name.

                return builder
            } catch (e: Exception) {
                throw RuntimeException(e)
            }
        }

        /**
         * A client that trusts the custom CAs. With pinnedHosts, only for those hosts,
         * decided per handshake so redirects are covered (ReactNativeBlobUtilPinnedTrustManager);
         * every other host gets the system's trust, which is what the README documents
         * and what iOS does by returning NSURLSessionAuthChallengePerformDefaultHandling.
         */
        @JvmStatic
        fun getCustomCACertOkHttpClient(
            client: OkHttpClient,
            context: Context,
            certNames: List<String?>,
            trustSystemCerts: Boolean,
            pinnedHosts: List<String?>? = null,
        ): OkHttpClient.Builder {
            try {
                val cf = CertificateFactory.getInstance("X.509")
                val keyStore = KeyStore.getInstance(KeyStore.getDefaultType())
                keyStore.load(null, null)

                if (trustSystemCerts) {
                    val defaultTmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm())
                    defaultTmf.init(null as KeyStore?)
                    var systemIndex = 0
                    for (tm in defaultTmf.trustManagers) {
                        if (tm is X509TrustManager) {
                            for (cert in tm.acceptedIssuers) {
                                keyStore.setCertificateEntry("system_${systemIndex++}", cert)
                            }
                        }
                    }
                }

                var loaded = 0
                for (certName in certNames) {
                    val cert = loadCertificateFromResources(context, cf, certName)
                    if (cert != null) {
                        keyStore.setCertificateEntry(certName, cert)
                        loaded++
                    }
                }

                // Fail closed. Without this the keystore is empty, every connection fails
                // trust evaluation, and the developer sees a generic handshake error rather
                // than the actual problem: the certificate is not in res/raw under that name.
                if (loaded == 0) {
                    throw IllegalStateException("customCACerts: none of $certNames could be loaded from res/raw")
                }

                val tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm())
                tmf.init(keyStore)

                val caTrustManager = tmf.trustManagers.firstOrNull { it is X509TrustManager } as X509TrustManager?
                    ?: throw IllegalStateException("No X509TrustManager found")
                val customTrustManager: X509TrustManager = if (pinnedHosts.isNullOrEmpty()) {
                    caTrustManager
                } else {
                    val systemTmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm())
                    systemTmf.init(null as KeyStore?)
                    val systemTrustManager = systemTmf.trustManagers.firstOrNull { it is X509TrustManager } as X509TrustManager?
                        ?: throw IllegalStateException("No system X509TrustManager found")
                    ReactNativeBlobUtilPinnedTrustManager(caTrustManager, systemTrustManager, pinnedHosts)
                }

                val sslContext = SSLContext.getInstance("TLS")
                sslContext.init(null, arrayOf<TrustManager>(customTrustManager), SecureRandom())

                val builder = client.newBuilder()
                builder.sslSocketFactory(sslContext.socketFactory, customTrustManager)

                // No hostnameVerifier override: the default verifier still applies, so a
                // certificate is only accepted for the names it actually carries.

                return builder
            } catch (e: Exception) {
                throw RuntimeException("Failed to configure custom CA certificates", e)
            }
        }

        private fun loadCertificateFromResources(context: Context, cf: CertificateFactory, certName: String?): Certificate? {
            var resId = context.resources.getIdentifier(certName, "raw", context.packageName)
            if (resId == 0) {
                for (suffix in arrayOf("_pem", "_der", "_cer")) {
                    resId = context.resources.getIdentifier(certName + suffix, "raw", context.packageName)
                    if (resId != 0) break
                }
            }
            if (resId == 0) return null

            try {
                context.resources.openRawResource(resId).use { return cf.generateCertificate(it) }
            } catch (e: Exception) {
                // DER parsing failed — try PEM-to-DER conversion
            }

            try {
                context.resources.openRawResource(resId).use {
                    val derBytes = pemToDer(it)
                    if (derBytes != null) {
                        return cf.generateCertificate(ByteArrayInputStream(derBytes))
                    }
                }
            } catch (e: Exception) {
                // PEM conversion also failed
            }

            return null
        }

        private fun pemToDer(input: InputStream): ByteArray? {
            return try {
                val reader = BufferedReader(InputStreamReader(input))
                val base64 = StringBuilder()
                while (true) {
                    val line = reader.readLine() ?: break
                    if (line.startsWith("-----")) continue
                    // Java's String.trim: only characters up to ' ', unlike Kotlin's trim().
                    base64.append(line.trim { it <= ' ' })
                }
                Base64.decode(base64.toString(), Base64.DEFAULT)
            } catch (e: Exception) {
                null
            }
        }

        /**
         * String to byte converter method
         *
         * @param data     Raw data in string format
         * @param encoding Decoder name
         * @return Converted data byte array
         */
        @JvmStatic
        private val BASE64_ALPHABET = Regex("[A-Za-z0-9+/=\\s]*")

        fun stringToBytes(data: String, encoding: String): ByteArray {
            if (encoding.equals("ascii", ignoreCase = true)) {
                return data.toByteArray(Charset.forName("US-ASCII"))
            } else if (encoding.lowercase(Locale.ROOT).contains("base64")) {
                // Android's decoder skips characters outside the alphabet, so
                // "@@@@" decoded to nothing and a write of it succeeded with 0 bytes.
                // iOS rejects such input; so does Android now.
                if (!BASE64_ALPHABET.matches(data)) {
                    throw IllegalArgumentException("Invalid base64 data")
                }
                return Base64.decode(data, Base64.NO_WRAP)
            } else if (encoding.equals("utf8", ignoreCase = true)) {
                return data.toByteArray(Charset.forName("UTF-8"))
            }
            return data.toByteArray(Charset.forName("US-ASCII"))
        }

        /**
         * The file path a string names: a plain path as it is, a file:// URI without
         * its scheme, a bundle-assets:// URI unchanged. Any other URI, content://
         * above all, is not a file path and yields null; callers open it through
         * ContentResolver (ReactNativeBlobUtilContent). Resolving a content URI to
         * a path let a URI shared by another app name this app's private files.
         *
         * @param path URI string.
         * @return The path, or null for a URI that is not one.
         */
        @JvmStatic
        fun normalizePath(path: String?): String? {
            if (path == null) return null
            if (!SCHEME.matches(path)) return path
            if (path.startsWith("file://")) {
                return path.replace("file://", "")
            }

            return if (path.startsWith(ReactNativeBlobUtilConst.FILE_PREFIX_BUNDLE_ASSET)) path else null
        }

        @JvmStatic
        fun isAsset(path: String?): Boolean =
            path != null && path.startsWith(ReactNativeBlobUtilConst.FILE_PREFIX_BUNDLE_ASSET)

        @JvmStatic
        fun isContentUri(path: String?): Boolean =
            path != null && path.startsWith(ReactNativeBlobUtilConst.FILE_PREFIX_CONTENT)
    }
}
